text/1025-branches.rst

```diff
@@ -0,0 +1,181 @@
+::
+
+    Status: Draft
+    Type: Feature
+    Created: 2024-01-11
+    Authors: Victor Petrovykh <victor@edgedb.com>
+
+
+===========================
+RFC 1025: Database Branches
+===========================
+
+Abstract
+========
+
+We propose to reframe the current concept of ``database`` in EdgeDB as
+``branch``. This avoids some of the confusion in current terminology and
+provides a clearer guideline for using EdgeDB in development environment.
+Functionally a ``branch`` will be the same as the current ``database``, but
+with some additional options and tooling.
+
+
+Motivation
+==========
+
+The current nomenclature of ``database`` and the corresponding commands are
+confusing to many users. A running EdgeDB server represents an *instance*
+while each instance can have multiple *databases* in it. Since the databases
+all run by the same server they all have to use the same underlying EdgeDB
+version. These qualities make them impractical for usage across multiple
+projects as the projects would have to share resources and perform EdgeDB
+upgrades together. We specifically recommend to have a dedicated instance for
+each project. This leaves databases as a feature without a clear usage.
+
+Instead we want to enhance the developer experience while using multiple
+source control branches. These multiple databases could map well onto
+corresponding source branches, representing slightly different version of the
+data and schema. Thus there is a good reason to rename the concept of
+*database* to *branch* and provide additional branch-handling tools. With this
+change it is also now no longer necessary to have a semantic distinction
+between the terms *instance* and *database* which may be natural for some
+users to think of as interchangeable.
+
+
+Branch DDL Commands
+===================
+
+There are several new DDL commands introduced by this proposal:
+
+1) ``create empty branch <newbranch>``
+
+   The most basic command creates a new branch with an empty schema
+   (exactly what ``create database`` command currently does).
+
+2) ``create branch <newbranch> [ from <oldbranch> ]``
+
```

Comment:
I guess this would be translated into `create branch <nb> from <ob>@state`.

In development, one regular worfklow is to checkout to an older commit, create a new branch and start from there. I guess in that case, it will pick the branch migration state from the schema file? or how would that work?

Reply:
Database branches do have their "current state", so it would be using that state.

A checkout of arbitrary state would be possible to implement, but would be slow, as we would have to create an empty branch and apply all needed migrations one-by-one.

There would be no way of keeping any data contents.